src/extensions/common/workspaceReducer.ts

```diff
@@ -0,0 +1,159 @@
+    recordingIds: string[]
+}
+
+type AddSortingsWorkspaceAction = {
+    type: 'ADD_SORTING'
+    sorting: Sorting
+}
+
+type DeleteSortingsWorkspaceAction = {
+    type: 'DELETE_SORTINGS'
+    sortingIds: string[]
+}
+
+type DeleteSortingsForRecordingsWorkspaceAction = {
+    type: 'DELETE_SORTINGS_FOR_RECORDINGS'
+    recordingIds: string[]
+}
+
+export interface AddUnitLabelWorkspaceAction {
+    type: 'ADD_UNIT_LABEL'
+    sortingId: string
+    unitId: number
+    label: string
+}
+
+export interface RemoveUnitLabelWorkspaceAction {
+    type: 'REMOVE_UNIT_LABEL'
+    sortingId: string
+    unitId: number
+    label: string
+}
+
+export interface MergeUnitsAction {
+    type: 'MERGE_UNITS'
+    sortingId: string
+    unitIds: number[]
+}
+
+export interface UnmergeUnitsAction {
+    type: 'UNMERGE_UNITS'
+    sortingId: string
+    unitIds: number[]
+}
+
+export interface SetSortingCurationWorkspaceAction {
+    type: 'SET_CURATION'
+    curation: SortingCuration
+}
+
+export type SortingCurationWorkspaceAction = AddUnitLabelWorkspaceAction | RemoveUnitLabelWorkspaceAction | SetSortingCurationWorkspaceAction | MergeUnitsAction | UnmergeUnitsAction
+export type WorkspaceAction = AddRecordingWorkspaceAction | DeleteRecordingsWorkspaceAction | AddSortingsWorkspaceAction | DeleteSortingsWorkspaceAction | DeleteSortingsForRecordingsWorkspaceAction | SortingCurationWorkspaceAction
+
+export const sortingCurationReducer = (state: SortingCuration, action: SortingCurationWorkspaceAction): SortingCuration => {
+    if (action.type === 'SET_CURATION') {
+        return action.curation
+    }
+    else if (action.type === 'ADD_UNIT_LABEL') {
+        const uid = action.unitId + ''
+        const labels = (state.labelsByUnit || {})[uid] || []
+        if (!labels.includes(action.label)) {
+            return {
+                ...state,
+                labelsByUnit: {
+                    ...state.labelsByUnit,
+                    [uid]: [...labels, action.label].sort()
+                }
+            }
+        }
+        else return state
+    }
+    else if (action.type === 'REMOVE_UNIT_LABEL') {
+        const uid = action.unitId + ''
+        const labels = (state.labelsByUnit || {})[uid] || []
+        if (labels.includes(action.label)) {
+            return {
+                ...state,
+                labelsByUnit: {
+                    ...state.labelsByUnit,
+                    [uid]: labels.filter(l => (l !== action.label))
+                }
+            }
+        }
+        else return state
+    }
+    else if (action.type === 'MERGE_UNITS') {
+        return {
+            ...state,
+            mergeGroups: simplifyMergeGroups([...(state.mergeGroups || []), action.unitIds])
+        }
+    }
+    else if (action.type === 'UNMERGE_UNITS') {
+        return {
+            ...state,
+            mergeGroups: simplifyMergeGroups((state.mergeGroups || []).map(g => (g.filter(x => (!action.unitIds.includes(x))))))
```

Comment:
Does this break if you have merge groups that consist of multiple units and you only remove one from them? Suppose a merge group is unit a merged with unit b merged with unit d. Now you remove unit d. Will units a & d still be merged, or will simplifying the merge groups wind up removing any merge that contains unit d?

(We may have been over this already, I feel like we discussed it before...)

Reply:
Unmerging is an operation on elements, not groups. So yes, if you unmerge d it gets removed from all the merge groups. This is okay, because generally the merge groups should be disjoint. In fact, we may want to enforce that merge groups are disjoint.